ARCHITECTURE.md

```diff
@@ -0,0 +1,133 @@
+
+In the context of this library, a request is an API to execute some OIDC-related operation. Handlers will determine whether they can handle the request based on the options specified by the user when making the call.
+
+#### Login
+
+Logging in is the operation that goes from the Client to the Issuer. It may result in a redirection of the user to the Issuer, but that is not always the case (in the `-node` module in particular). Handlers for the login operation are located in `packages/*/src/login/oidc/oidcHandlers/*Handler.ts`.
+
+#### Incoming redirect
+
+OIDC is based on redirection of the user's browser between the Client and the Issuer. The incoming redirect happens after the Resource Owner has authenticated to the Issuer, and the Issuer redirects them to the Client's `redirect_uri` with some relevant information in the query params. Handlers for the incoming redirect are located in `packages/*/src/login/oidc/redirectHandler/*Handler.ts`.
+
+### Dependency injection
+
+An important architectural component of this library is dependency injection, implemented here using [TSyringe](https://github.com/Microsoft/tsyringe). Dependencies are declared in `packages/*/src/dependencies.ts`.
+
+#### Declaring order
+
+Something important to realize is that the order in which the dependencies are declared (for a given container) matters. Let's have a look at some code to make things clearer.
+
+```
+container.register<IOidcHandler>("browser:oidcHandler", {
+  useClass: AggregateOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: RefreshTokenOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: AuthorizationCodeWithPkceOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: ClientCredentialsOidcHandler,
+});
+```
+
+Here, `AggregateOidcHandler` is the handler aggregator (as defined in the Handler Pattern section), and `RefreshTokenOidcHandler`, `AuthorizationCodeWithPkceOidcHandler` and `ClientCredentialsOidcHandler` are its underlying handlers. When receiving a request, `AggregateOidcHandler` will first call to its instance of `RefreshTokenOidcHandler` to check if it can handle it. If so, the instance of `RefreshTokenOidcHandler` will handle the request, and the instances of `AuthorizationCodeWithPkceOidcHandler` or `ClientCredentialsOidcHandler` will not be called. This means that it is important to declare the dependencies from the most specialized to the most generic, because if a fallback handler that can handle all requests is declared first, the other more specialized handlers will not be called.
```

Comment:
Wait, if the order in which the Handlers are added to `container` determines which gets called, then why does `AggregateOidcHandler` need to get called first which I'd _also_ assume to have a list of Handlers to determine which gets called? (In other words, what does `AggregateOidcHandler` do here that `container` doesn't?)

Reply:
Note that the label for the containers of the aggregator and the underlying handlers differ (respectively `browser:oidcHandler` and `browser:oidcHandlers`, with an 's'). In this case, the order in which the dependencies are declared is irrelevant, because they do not relate to the same container. The Aggregator implements the class from `packages/core/src/util/handlerPattern/AggregateHandler.ts`, and uses the `@injectAll` annotation to receive all the handlers registered to a given container. If this comment clarifies things, I'll add it to the documentation.

Reply:
Aha! Yes, that's definitely worth pointing out - at least the `s` part.

Reply:
4a9e1d4bc4eb45508cf141ac1ac1318ee824fba8